### PR TITLE
Add tests

### LINKED ISSUE #96: match_name(): Add test cases with explicit, known inputs and outputs
We need to confirm that `match_name()` outputs the expected values. At least we should capture the some simple cases and some edge cases (@jdhoffa, can you help create such cases?). This will also help us refactor the code with confidence that out tests will let us know when we break something.

A good way to do this is with inputs and outputs that are as minimal as possible. For example:

https://github.com/2DegreesInvesting/r2dii.match/blob/7d2e1959ac307acd43647210b4464f432131e0a4/tests/testthat/test-match_name.R#L44

To help create minimal, text-based data we now have the helpers [`mini_lbk()` and `mini_ald()`](https://github.com/2DegreesInvesting/r2dii.match/blob/master/R/mini_lbk.R).

### PR DESCRIPTION
**Pull request by maurolepore**
Set up PR to push new tests

Closes #96
Thanks @jdhoffa

### CODE REVIEW

**Comment by maurolepore** [comment]
@jdhoffa, I added a few cases at the top of test-match_name.R. If you can please review and/or add test cases you think we need. 

In the commits messages above you'll see I mentions you a few times. Where i would like your input the most is in deciding what to do when invalid sector or codes result in silent mismatches (a warning would only happen when there is no match a all). See https://github.com/2DegreesInvesting/r2dii.match/pull/99/commits/12df8f59b95cf4ac363e167722eaf0a085c9f937.

**Comment by maurolepore** [comment]
Also we should add case-tests for `prioritize()`. I developed it on my own and would benefit from you conceptual input (apparently here is a bug https://github.com/2DegreesInvesting/r2dii.match/issues/101).

**Comment by jdhoffa** [comment]
@maurolepore just had a look at 12df8f5, and all the tests in this commit seem good and make sense to me!
What did you want me to review re: silent issues?

As far as I can tell, match_name:
- [ ] Passes an error if the sector code_system provided is bad/ not in our data
- [ ] Properly matches 1 lbk element to 2 corresponding ald elements in different sectors
- [ ] Outputs a null result if all input is `NA`
- [ ] Can handle `NA` at crucial levels 

What was the silent issue you wanted my advice with? 

Beyond that, I had a look at each test added in this PR and have added my changes as necessary

File: tests/testthat/test-match_name.R [hunk 1]
```diff
@@ -168,11 +168,14 @@ test_that("match_name with missmatching sector_classification yields no match",
...
 
-  # Here they do NOT match and we expect no match
+  # Here they do NOT match and we expect no match if by_sector=T
   sector_not_power <- "coal"
   ald_mini$sector <- sector_not_power
   out <- expect_warning(match_name(lbk_mini1, ald_mini), "no match")
   expect_equal(nrow(out), 0L)
+
+  # If by_sector=F, we expect match
+  expect_equal(nrow(match_name(lbk_mini1,ald_mini, by_sector=F)), 1L)
```
Comment (maurolepore) [inline tests/testthat/test-match_name.R]: Nice work!

**Comment by maurolepore** [comment]
RE silent issues.

Compare:

### Good

The error message clearly indicates the problem and hins a solution:

``` r
suppressPackageStartupMessages(library(dplyr))
library(r2dii.dataraw)
#> Loading required package: r2dii.utils
library(r2dii.match)

bad_system <- "bad system"

lbk_mini1 <- tibble::tibble(
  sector_classification_system = bad_system,
  id_ultimate_parent = "UP15",
  name_ultimate_parent = "Alpine Knits India Pvt. Limited",
  id_direct_loantaker = "C294",
  name_direct_loantaker = "Yuamen Xinneng Thermal Power Co Ltd",
  sector_classification_direct_loantaker = 3511
)

ald_mini <- tibble::tibble(
  name_company = "alpine knits india pvt. limited",
  sector = "power",
  alias_ald = "alpineknitsindiapvt ltd"
)

match_name(lbk_mini1, ald_mini)
#> Error: At least one loan must use 2dii's sector code system.
#> Are all of your loans classified as in 2dii's database?
```

### Bad

The warning refers to a totally unrelated problem. The problem is that the value of `sector_classification_direct_loantaker` is bad. The warning is because this particular dataset has only one non-matching row, so the output has 0 match. But had I used a different dataset where some other name has a match, then we would get no warning at all.


``` r
suppressPackageStartupMessages(library(dplyr))
library(r2dii.dataraw)
#> Loading required package: r2dii.utils
library(r2dii.match)

code_does_not_exist <- -9999L

lbk_mini1 <- tibble::tibble(
  sector_classification_direct_loantaker = code_does_not_exist,
  id_ultimate_parent = "UP15",
  name_ultimate_parent = "Alpine Knits India Pvt. Limited",
  id_direct_loantaker = "C294",
  name_direct_loantaker = "Yuamen Xinneng Thermal Power Co Ltd",
  sector_classification_system = "NACE"
)

ald_mini <- tibble::tibble(
  name_company = "alpine knits india pvt. limited",
  sector = "power",
  alias_ald = "alpineknitsindiapvt ltd"
)

match_name(lbk_mini1, ald_mini)
#> Warning: Found no match.
#> # A tibble: 0 x 16
#> # … with 16 variables: sector_classification_direct_loantaker <int>,
#> #   id_ultimate_parent <chr>, name_ultimate_parent <chr>,
#> #   id_direct_loantaker <chr>, name_direct_loantaker <chr>,
#> #   sector_classification_system <chr>, id <chr>, level <chr>, sector <chr>,
#> #   sector_ald <chr>, name <chr>, name_ald <chr>, alias <chr>, alias_ald <chr>,
#> #   score <dbl>, source <chr>
```


This kind of  "silent" issue happens at least with:

* Bad code.
* Bad sector classification.
* Missmatch between code and sector.

@jdhoffa, do you think we need to check that all codes and sector classifications given in loanbook are valid? I believe  valid values should come from here (please correct me if I'm wrong):

``` r
library(r2dii.dataraw)
#> Loading required package: r2dii.utils
r2dii.match:::sector_clasiffication_df()
#> # A tibble: 2,600 x 4
#>    sector       borderline code  code_system
#>    <chr>        <lgl>      <chr> <chr>      
#>  1 not in scope FALSE      A     ISIC       
#>  2 not in scope FALSE      1     ISIC       
#>  3 not in scope FALSE      11    ISIC       
#>  4 not in scope FALSE      111   ISIC       
#>  5 not in scope FALSE      112   ISIC       
#>  6 not in scope FALSE      113   ISIC       
#>  7 not in scope FALSE      114   ISIC       
#>  8 not in scope FALSE      115   ISIC       
#>  9 not in scope FALSE      116   ISIC       
#> 10 not in scope FALSE      119   ISIC       
#> # … with 2,590 more rows
```

<sup>Created on 2020-01-22 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

File: tests/testthat/test-match_name.R (hunk 1 above)
Comment (jdhoffa) [reply tests/testthat/test-match_name.R]: 💃

**Comment by maurolepore** [comment]
Thanks @jdhoffa. The problem about silent issues persists, except when the problem is in the column `sector_classification_system`. Should I write assertions to stop? warn? do nothing?

See this reprex, now with the help of the new internal `fake_lbk()` and fake_ald()` which allow us to invalidate values in tests in a succinct way:

``` r
library(r2dii.dataraw)
#> Loading required package: r2dii.utils
library(r2dii.match)



r2dii.match:::fake_lbk
#> function(sector_classification_system = NULL,
#>                           id_ultimate_parent = NULL,
#>                           name_ultimate_parent = NULL,
#>                           id_direct_loantaker = NULL,
#>                           name_direct_loantaker = NULL,
#>                           sector_classification_direct_loantaker = NULL) {
#>   tibble::tibble(
#>     sector_classification_system = sector_classification_system %||% "NACE",
#>     id_ultimate_parent = id_ultimate_parent %||% "UP15",
#>     name_ultimate_parent =
#>       name_ultimate_parent %||% "Alpine Knits India Pvt. Limited",
#>     id_direct_loantaker = id_direct_loantaker %||% "C294",
#>     name_direct_loantaker =
#>       name_direct_loantaker %||% "Yuamen Xinneng Thermal Power Co Ltd",
#>     sector_classification_direct_loantaker =
#>       sector_classification_direct_loantaker %||% 3511
#>   )
#> }
#> <bytecode: 0x000000001af92aa0>
#> <environment: namespace:r2dii.match>

r2dii.match:::fake_ald
#> function(name_company = NULL,
#>                           sector = NULL,
#>                           alias_ald = NULL) {
#>   tibble::tibble(
#>     name_company = name_company %||% "alpine knits india pvt. limited",
#>     sector = sector %||% "power",
#>     alias_ald = alias_ald %||% "alpineknitsindiapvt ltd"
#>   )
#> 
#> }
#> <bytecode: 0x000000001b03ca10>
#> <environment: namespace:r2dii.match>



match_name(
  r2dii.match:::fake_lbk("bad"),
  r2dii.match:::fake_ald()
)
#> Error: At least one loan must use 2dii's sector code systems:
#> ISIC, NACE, NAICS
#> Are all of your loans classified as in 2dii's database?

match_name(
  r2dii.match:::fake_lbk( , "bad"),
  r2dii.match:::fake_ald()
)
#> # A tibble: 1 x 16
#>   sector_classifi… id_ultimate_par… name_ultimate_p… id_direct_loant…
#>   <chr>            <chr>            <chr>            <chr>           
#> 1 NACE             bad              Alpine Knits In… C294            
#> # … with 12 more variables: name_direct_loantaker <chr>,
#> #   sector_classification_direct_loantaker <dbl>, id <chr>, level <chr>,
#> #   sector <chr>, sector_ald <chr>, name <chr>, name_ald <chr>, alias <chr>,
#> #   alias_ald <chr>, score <dbl>, source <chr>

match_name(
  r2dii.match:::fake_lbk( , , "bad"),
  r2dii.match:::fake_ald()
)
#> Warning: Found no match.
#> # A tibble: 0 x 16
#> # … with 16 variables: sector_classification_system <chr>,
#> #   id_ultimate_parent <chr>, name_ultimate_parent <chr>,
#> #   id_direct_loantaker <chr>, name_direct_loantaker <chr>,
#> #   sector_classification_direct_loantaker <dbl>, id <chr>, level <chr>,
#> #   sector <chr>, sector_ald <chr>, name <chr>, name_ald <chr>, alias <chr>,
#> #   alias_ald <chr>, score <dbl>, source <chr>

match_name(
  r2dii.match:::fake_lbk( , , , "bad"),
  r2dii.match:::fake_ald()
)
#> # A tibble: 1 x 16
#>   sector_classifi… id_ultimate_par… name_ultimate_p… id_direct_loant…
#>   <chr>            <chr>            <chr>            <chr>           
#> 1 NACE             UP15             Alpine Knits In… bad             
#> # … with 12 more variables: name_direct_loantaker <chr>,
#> #   sector_classification_direct_loantaker <dbl>, id <chr>, level <chr>,
#> #   sector <chr>, sector_ald <chr>, name <chr>, name_ald <chr>, alias <chr>,
#> #   alias_ald <chr>, score <dbl>, source <chr>

match_name(
  r2dii.match:::fake_lbk( , , , , "bad"),
  r2dii.match:::fake_ald()
)
#> # A tibble: 1 x 16
#>   sector_classifi… id_ultimate_par… name_ultimate_p… id_direct_loant…
#>   <chr>            <chr>            <chr>            <chr>           
#> 1 NACE             UP15             Alpine Knits In… C294            
#> # … with 12 more variables: name_direct_loantaker <chr>,
#> #   sector_classification_direct_loantaker <dbl>, id <chr>, level <chr>,
#> #   sector <chr>, sector_ald <chr>, name <chr>, name_ald <chr>, alias <chr>,
#> #   alias_ald <chr>, score <dbl>, source <chr>

match_name(
  r2dii.match:::fake_lbk( , , , , , "bad"),
  r2dii.match:::fake_ald()
)
#> Warning: Found no match.
#> # A tibble: 0 x 16
#> # … with 16 variables: sector_classification_system <chr>,
#> #   id_ultimate_parent <chr>, name_ultimate_parent <chr>,
#> #   id_direct_loantaker <chr>, name_direct_loantaker <chr>,
#> #   sector_classification_direct_loantaker <chr>, id <chr>, level <chr>,
#> #   sector <chr>, sector_ald <chr>, name <chr>, name_ald <chr>, alias <chr>,
#> #   alias_ald <chr>, score <dbl>, source <chr>
```

<sup>Created on 2020-01-23 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by jdhoffa** [comment]
@maurolepore apologies Mauro, I see the point now. 
- `id_ultimate_parent`: I don't think there needs to be an error/ warning for this. The internal ID's we generate are a certain format, but in principle I think other formats of IDs should be fine, so long as they are unique. I'd rather not be too forceful here if possible. 
- `id_direct_loantaker`: ibid
- The `name` columns are fine. A "bad" name is just a name that doesn't match to anything, which is expected to happen and should pass silently

The only place I can think to add a warning/ error message is if `sector_classification_direct_loantaker` is not in `(power, oil&gas, coal, automotive, cement, shipping, steel, aviation)`. However, if these sectors are "good" or "bad" is decided by `r2dii.dataraw` and not by the user (so throwing an error seems a bit silly, because the user can't do anything about it. it would be a bug on our part).
- I can write a test in r2dii.dataraw to ensure that every *_classification file only has those sectors?

**Comment by maurolepore** [comment]
The commits above refactor tests for `prioritize()`, mostly to DRY -- as an investment in maintainability.

**Comment by maurolepore** [comment]
@jdhoffa 

RE:

> The only place I can think to add a warning/ error message is if sector_classification_direct_loantaker is not in (power, oil&gas, coal, automotive, cement, shipping, steel, aviation). However, if these sectors are [...] "bad" [...] the user can't do anything about it.

Is it impossible for the bad value of `sector_classification_direct_loantaker` would come from the user's loanbook?

``` r
suppressWarnings(library(dplyr))
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(r2dii.match)
library(r2dii.dataraw)
#> Loading required package: r2dii.utils

user_loanbook <- r2dii.match:::fake_lbk(
  sector_classification_direct_loantaker = c("oil&gas", "oil and gas")
)

user_loanbook %>% 
  select(matches("sector.*loantaker"), everything())
#> # A tibble: 2 x 6
#>   sector_classifi… sector_classifi… id_ultimate_par… name_ultimate_p…
#>   <chr>            <chr>            <chr>            <chr>           
#> 1 oil&gas          NACE             UP15             Alpine Knits In…
#> 2 oil and gas      NACE             UP15             Alpine Knits In…
#> # … with 2 more variables: id_direct_loantaker <chr>,
#> #   name_direct_loantaker <chr>

match_name(user_loanbook, r2dii.match:::fake_ald())
#> Warning: Found no match.
#> # A tibble: 0 x 16
#> # … with 16 variables: sector_classification_system <chr>,
#> #   id_ultimate_parent <chr>, name_ultimate_parent <chr>,
#> #   id_direct_loantaker <chr>, name_direct_loantaker <chr>,
#> #   sector_classification_direct_loantaker <chr>, id <chr>, level <chr>,
#> #   sector <chr>, sector_ald <chr>, name <chr>, name_ald <chr>, alias <chr>,
#> #   alias_ald <chr>, score <dbl>, source <chr>
```

<sup>Created on 2020-01-24 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by maurolepore** [comment]
RE:

> I can write a test in r2dii.dataraw to ensure that every *_classification file only has those sectors?
-- @jdhoffa,

**Comment by jdhoffa** [comment]
Ah you're correct! So in principle it is impossible for the sector to be wrong, since it is defined by a left_join, however it could arise from the overwrite file, when a sector is overwritten.

**Comment by maurolepore** [comment]
Thanks! There are still some details I don't understand but I'll discuss separately. This PR has already useful work and the details can wait.
